MD Tab Update

podraco opened the pull request:
### Description of Changes
this is a fix for some problems that surged from the last pull request, as well add some improvements and upgrades to the class.

Possible fix for #770 
Possible fix for #874 
Possible fix for #878 
Possible fix for #880 
Possible fix for #885

This is a big change for the class, it has been proven with every example code provided by the documentation and kitchen sink.

This updates the documentation for the new implemented methods.

Between the fixes there are:
- simplified the `remove_widget ` function from `MDTabs`
- `remove_widget ` now removes binds created on `add_widget` for new tabs.
- max width of every tabt's title label is set to `"360dp"`
- both kv and python instance creation reviewed. (small differences that affected its behavior on new widgets)
- added exception raises if something is not filled correctly or found.
- Added more specific Exceptions to allow developers know what exactly went wrong.
- Added basic `MDCarousel` Documentation.
- `MDTabsBase.text` is now deprecated, still has backward compatibility, but the use `tab_label_text` is now enforced and the use of this property will show a message in the kivy logger.
- Fixed reference bug in which, if the tab with the reference was pressed, the selected tab would be missing its state.
- Tab_bar now behaves on the content's width. 
If the width of the tab's inside is bigger than the scroll view, the `tab_bar` will be set to scrollable, the padding of the layout will be set to `["52dp", 0]` and allow scroll, if the width is lesser or equal to the scrollview width, the scroll then is locked, and the canvas padding is set to `[0,0]`,  `ScrollView.do_scroll_x` is set to `False` in this scenario.
- Some duplicated code was removed
- Some duplicated movements were removed, since some steps were already done by the children widget or by another event.    
- Fixed some small recursions.

Adds:
- tab_label_text namespace will store the raw tab's title label text. Replaces `MDTabsBase.text` property.
-  Tabs.switch_to is now more flexible, you can toggle between text, Tab_objects and if it's a string, search by its icon, title or raw text.
- Any text parsed to the Title property will be set with upper(). (this is to maintain every tab title clear)
- added title_is_capital, if the value is True, it will update the self. Text to be all capitals.

podraco committed:
removed unused kv code

podraco committed:
added some iteration between scroll behavior and fixed behavior

this applies to the tab scroll view. if the tabs width is minor or equal 
to the scroll view, the scroll will be fixed.
If not, will be free to scroll and add a padding of 52 dp horizontally 
as the spec requires.

podraco committed:
Fixed small details of the label configurations.

podraco committed:
removed forced boldness of the label

podraco committed:
Fix for allow_stretch bug

added `fixed_tab_label_width`
This will set the deffault size of the tab when `allow_stretch` is 
`False`.

podraco committed:
changed unused code because of the font_style

podraco committed:
removed unused code

podraco committed:
fixed max_width issue, set to 150 instead 360

podraco committed:
added unbinding methods to remove widgets

podraco committed:
added font_style capability to the MDTab Base

podraco committed:
Merge pull request #1 from kivymd/master

Update private with master fork the latest changes

podraco committed:
Fix for tabs unable to set title on kvlang

podraco committed:
Fixed wrongly width scrollable area.

this will fix the issue in wich if the tab was in scroll mode and a 
children is removed, the scrollview never updated to paddin 0 if the 
total width of the children was less than the scroll width, because the 
container never removed the padding.

podraco committed:
Fixed issue in wich if a tab was removed, the canvas reset to index 0

This improve the efficiency of the MDTab remove_widget. giving it 
flexibility to accept removal of either MDTabsLabel or a MDTabsBase 
instance.

Should we consider clear_widget?

podraco committed:
Fixed issue in wich removing a tab would set the carousle index to 0

podraco committed:
Black run

podraco committed:
Fixed issue in wich the tabs list was references backwards.

this is caused because the children list works inverted to a normal 
list, the lastest element added to the list is the first on the children 
list.

podraco committed:
Fixed Tabs documentation

podraco committed:
added MDCarousel documentation

podraco committed:
deprecated value added to tabsbase text, use tab_label_text instead

podraco committed:
Fixed issues with the new `tab_label_text` property compatibility.

podraco committed:
Docs update

podraco committed:
bug fix in for referenced text

The issue was that when a ref text was clicked on, the tab label was set 
to down and if the call back removed the widget, the tab_bar had not 
element with state="down"

podraco committed:
Fixed Docs to simplify work and examples.

podraco committed:
Docs Update

podraco committed:
updated switch_tab to allow new features.

by deffect the fucntion will search with the tab_label_text if the 
argument (name_tab) is a String, if it's a MDTabsBase. the tab will be 
chagned to it instead of iterating between all tabs.

podraco committed:
Pre-commit run

podraco committed:
avoids calling a second update event

podraco committed:
Doc update
fixed missing tabs, if the value doesnt exist, it will throw a 
ValueError exception.

podraco committed:
Docs update.

podraco committed:
Docs Update

podraco committed:
updated ussage for the lastest changes.

podraco committed:
run Pre-commit

podraco committed:
Buxfix and improvement.

recursion fix

removed unnecesary size management, migrated from MDtabsbase.text to 
MDtabsbase.tab_label_text.

podraco commented:
i have done multiple tests with different codes that where available on the issues reports and on the documentations and demos. i believe this changes works and fulfill the normal behavior of any kivyMD and comply with the Material IO standard.

If there's something that needs to be fixed, ill work on it as soon as possible as I can.

podraco commented:
I'm waiting response from a collaborator, please don't approve merge yet.

runette reviewed (commented):
As a general comment - I still have the problem that we are fundamentally changing the effect that `allow_stretch` has. This means that ALL previous apps that used `allow_stretch` MUST be changed and this does not even seem to be documented.

BEFORE (i.e. before about 2 weeks ago):

`allow_stretch=True` created the behavior described in material.io as "Fixed Tabs" - i.e. if possible it stretched the Tabs wider than the label text so that they filled the TabBar.

AFTER (i.e. now) :

`allow_stretch=True` means that the Tab will expand to the width of the text (i.e. pretty much what `allow_stretch = False` meant before.

I find it unfortunate and unnecessary to fundamentally reverse the meaning of an existing API element like this. It would be much better to deprecate the old behavior and create a new attribute to trigger the new behavior.

runette commented on kivymd/uix/tab.py:
```diff
@@ -720,40 +788,81 @@ class MDTabsBase(Widget):
+
+    def _set_label_font_style(self, value):
+        if self.tab_label:
+            if value in theme_font_styles:
```
This shold be compared to the actual and current font_style and not the imported one. Since TabLabel inherits from MDLabel it also inherits from ThemeableBehavior so this should be 

`if value in self.tab_label.theme_cls.font_styles`

podraco replied:
The documentation stated that was the proper use to register and read a new font. Let me test it in a different demo to see if that works, if not, a new issue must be opened

runette replied:
As is usual, the problem is actually in how you state the question!

In this code - you should not be trying to register or read a *new* font. You are trying to read the fonts and actually font_styles that *have been* registered with the app.

If you use the current code - you will *only* get the standard fonts provided out of the box. If the app developer has followed the procedure as documented to add a new font to their app - it will be ignored by your code. This is obviously a bug in this code. I do not see the need for a new issue.

podraco replied:
The code only uses what's stored in memory.

runette replied:
That is misdirection. It is all stored in memory. The question is whether you are using the correct values in the correct memory.

runette commented on kivymd/uix/tab.py:
```diff
@@ -481,7 +520,7 @@ def switch_tab(self):
 from kivy.utils import boundary
 
-from kivymd.font_definitions import fonts
+from kivymd.font_definitions import fonts, theme_font_styles
```
As a general comment - components should use the ThemableBehavior to get the current font_definitions and not import them - since they may have been changed by the application designer in the App.

podraco replied:
Well, as far as i know, when a dic is imported. The `__import__` method only creates a link to the pointer, and add the value to the global environment rather to contain it in the file/lib. Because of this, you could alter the value in any file and every change should be shared. Yet, I'm not certain if the event dispatcher would call any event to it.

Also, as far as I know, comparing a value to a list were always faster than to a dict.

![Captura de pantalla de 2021-02-23 17-38-17](https://user-images.githubusercontent.com/5509325/108923182-fb67ed80-75fd-11eb-82a6-e337e8514d38.png)
![Captura de pantalla de 2021-02-23 17-37-22](https://user-images.githubusercontent.com/5509325/108923185-fc991a80-75fd-11eb-80dc-600bd63df2c4.png)
![Captura de pantalla de 2021-02-23 17-37-33](https://user-images.githubusercontent.com/5509325/108923190-fe62de00-75fd-11eb-9438-ccf5b8bdf083.png)
![Captura de pantalla de 2021-02-23 17-39-16](https://user-images.githubusercontent.com/5509325/108923249-163a6200-75fe-11eb-8725-e8788851924f.png)

podraco replied:
The last commit ensures parity between the pointer and the theme_cls object.

runette replied:
To me, you are adding complexity on top of complexity and adding a lot of new code and possible error modes for the relatively small performance gain for getting the list of values from a list rather than dict items. 

The class `ThemeManager` HAS an existing attribute that holds the current font styles. Why not just always drive from the current value of that attribute rather than require another value to be kept up to date?

If I was technical lead on this project I would say this solution adds to the error vectors of the project and makes the project less maintainable - and indeed would probably not give this change "readability' approval. But I think this will work.

runette commented on kivymd/uix/tab.py:
```diff
@@ -720,40 +788,81 @@ class MDTabsBase(Widget):
+        # ensures that the title is in capital letters.
         if self.title:
-            self.title = self.title.upper()
+            if self.title != self.title.upper():
```
PLEASE - do not hard code the label title to Upper case.

This is a deal-breaker. There is NO specification in material.io that says that Tab labels should be all caps.

While you might think it is good practice and I might agree with you - this MUST be a matter for the app designer.

podraco replied:
Well, the material io requieres that all labels must be clear, and. As far as I know, common practice to this is to use capital letters, since they have the most serif and use a bigger space to be drawn. But there's no problem, it can be removed

podraco replied:
yet, i just changed it to be optional.

runette replied:
> As far as I know, common practice to this is to use capital letters, since they have the most serif and use a bigger space to be drawn

Thanks for making it optional.

Hard coding "good practice" is really not a good practice. And I do not necessarily agree with you. I use san-serif fonts (and Roboto is also san-serif) so that the question of the size of the serif has no relevance.

And if you look at from the point of view of psychology - all-caps words are harder for humans to read because we actually read by pattern recognition on the word and not by reading the letters.

runette commented on kivymd/uix/tab.py:
```diff
@@ -1079,12 +1172,25 @@ class MDTabs(ThemableBehavior, SpecificBackgroundColorBehavior, AnchorLayout):
     and defaults to `True`.
     """
 
+    fixed_tab_label_width = NumericProperty("140dp")
```
See the general comment about the change in the meaning of `allow_stetch`

In addition - although this mechanism works and is broadly in line with material.io - I do miss the functionality to ask kivyMD to expand the Tabs to fill the screen. It was useful functionality and was also in line with material.io

podraco replied:
It is a useful attribute, but, that shouldn't be it's name, since hat you're requiring, it's size_hint_x, from the grid layout that contains all tabs, implementing something to do the same in cpython to normal python is a really bad practice, but, i could forward the size hint of the tab bar to the MDTabsBase. So you could still use something like.
```py
tab_hint_x=1
tab_hint_x=None
```

runette replied:
1. It is not a matter of whether you think the name is a good name or whether I think it is a good name - IT IS THE EXISTING API. That is what existing application - some of them things that people make money out of - have included in their applications. Not because they are stupid or bad programmers but because that is what the previous specification WAS.

2. I did not suggest anything about implementation - so I am not sure why you seem to think that I was proposing to change something from cython to python? 

3. I am not sure that I understand the proposed solution - not enough info - but if it involves adding NEW complexity to an already complex API then probably not.

I think what I would really would have liked is:

- `allow_stretch` doing exactly what it did before (because :: legacy) and deprecated (because :: not really MD and badly named), and

- A "tab mode" attribute that has values of  "Fixed", "Scrollable" and "Fixed-length" for the various modes in material.io (or something similar to this).

runette reviewed (changes requested):
Starting to look good.

I do see some fairly fundamental bugs - as listed below

runette commented:
General Comment :

When I run the new code in Kitchen Sink - there is a gap to the left of the first tab that I don't remember being there before. I don't have the time today to do a side-by-side analysis to confirm that. It does not look like a phantom tab because it cannot be selected.

<img width="912" alt="Screenshot 2021-02-23 at 11 15 05" src="https://user-images.githubusercontent.com/2239795/108838731-effbce80-75cb-11eb-850f-2995dd171c19.png">

runette commented:
General Comment :

When I run this new tab.py in my existing (released) application without other changes - I get this error message.

Throwing an error on legacy code that was working before could be a significant problem for the existing users of kivyMD.
```
builtins.ValueError: switch_tab:
name_tab not found in the tab list
search_by = 'text'
name_tab = 'Basic Details'
```

HeaTTheatR commented:
@runette  Didn't know the labels were so different on `Mac OS` :)

![Без имени](https://user-images.githubusercontent.com/16930280/108840475-78d43380-75e7-11eb-97da-2a80674c0ee7.png)

runette commented:
> @runette Didn't know the labels were so different on `Mac OS` :)

I am using a custom font - mostly to test the most extreme case but also because my app uses the corporate brand font.

podraco commented:
That gap is required for scrollable tabs, it's in the specs and it's supposed to be `"52dp"`
If the width of every children in the container is lessed than the scroll view width, the do_scroll_x is False, and the padding =[0,0], this behavior updates every time the width is updated or a tab is added/removed/updated.

podraco commented:
> General Comment :
> 
> When I run this new tab.py in my existing (released) application without other changes - I get this error message.
> 
> Throwing an error on legacy code that was working before could be a significant problem for the existing users of kivyMD.
> 
> ```
> builtins.ValueError: switch_tab:
> name_tab not found in the tab list
> search_by = 'text'
> name_tab = 'Basic Details'
> ```

That was my error, forgot to simplify the text than to use the property name as the default value

podraco committed:
missing deffault search value

podraco committed:
Added configuration property.
Increase flexibility to allow the developer decide if the tab label text 
should be or not Capital

podraco commented:
> General Comment :
> 
> When I run the new code in Kitchen Sink - there is a gap to the left of the first tab that I don't remember being there before. I don't have the time today to do a side-by-side analysis to confirm that. It does not look like a phantom tab because it cannot be selected.
> 
> <img alt="Screenshot 2021-02-23 at 11 15 05" width="912" src="https://user-images.githubusercontent.com/2239795/108838731-effbce80-75cb-11eb-850f-2995dd171c19.png">

its because of this.

![Captura de pantalla de 2021-02-23 15-14-06](https://user-images.githubusercontent.com/5509325/108909191-d9647000-75e9-11eb-8109-351458663642.png)

![Captura de pantalla de 2021-02-23 15-57-38](https://user-images.githubusercontent.com/5509325/108913674-e8e6b780-75ef-11eb-8b11-11ca0caa527c.png)

podraco committed:
This update keeps track of every font style change in the theme manager.

podraco committed:
run pre-commit

podraco committed:
cleaned binding calls.

podraco committed:
Added tab_hint_x as a Boolean value.

this will not break the max width of 360dp of each tab.

podraco committed:
Run pre-commit

podraco commented:
So, these are the latest changes, fix some recursion calls by events, and adds what I believe was the last needed feature.
I'll wait until it can be reviewed.

I suggest that we merge this until everyone involved approves this PR.

runette commented on kivymd/theming.py:
```diff
@@ -1052,7 +1053,18 @@ def __init__(self, **kwargs):
+            if style not in self.font_styles:
+                theme_font_styles.pop(num)
+        for style in self.font_styles.keys():
+            theme_font_styles.append(style)
```
See comment below.

To me - this looks like a complex solution to fix a problem that does not really exist.

podraco replied:
Both namespaces should keep the same keys, values, otherwise, why should it exists as separated elements?

runette commented:
I have no further technical issues with this PR - it is working for me both in my app an in kitchen garden.

My only remaining concern is the breaking change in the API (i.e. the change in meaning and usaga of `allow_stretch`).

I will leave that to @HeaTTheatR to decide.

runette commented:
BTW - just a procedural tip that you probably know already - if instead of 

"Possible fix for #770"
...

you had 

"fixes #770"
...

then merging the PR would automatically close the issue.

see https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

syleishere commented:
Just tested this again, allow_stretch still won't work without tab_hint_x: True

podraco commented:
Allow stretch is by defect `True`.
Allow stretch will allow the tab to stretch to the maximum size of the tab by the label content before wrapping it to 2 lines.
if `allow_stretch` is `False`, the tab won't change to the text texture, instead, it will use the fixed width property to set the tab's label width.

Meanwhile, `Tab_hint_x` will work outside the tab's texture size, and allow the tab to use a size hint based on the MDTabs widget width.
